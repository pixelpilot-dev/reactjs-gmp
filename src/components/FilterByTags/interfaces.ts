import { TGenresProps } from '../../core/types/TGenresProps';

export interface IFilterByTagsProps {
  options: TGenresProps;
}
