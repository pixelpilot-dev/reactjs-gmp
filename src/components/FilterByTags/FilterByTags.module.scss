.filter {
  display: flex;

  .button {
    @include resetbtn;
    font-size: 16px;
    line-height: em(20, 16);
    text-transform: uppercase;
    color: $white;
    position: relative;
    padding: 20px 0;
    margin-right: 30px;
    transition: color 0.3s linear;

    @include for-mobile {
      font-size: 14px;
      padding: 15px 0;
      margin-right: 10px;
    }

    &::after {
      content: '';
      display: block;
      position: absolute;
      bottom: -4px;
      left: 0;
      width: 100%;
      border-bottom: 4px solid $primary;
      opacity: 0;
      visibility: hidden;
    }

    &:hover {
      color: $primary;
    }

    &.isActive::after {
      opacity: 1;
      visibility: visible;
    }
  }
}
