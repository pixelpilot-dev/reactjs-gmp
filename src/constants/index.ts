import { range } from 'lodash';
import { getYear } from '../utils/date';

export const MONTH = [
  'January',
  'February',
  'March',
  'April',
  'May',
  'June',
  'July',
  'August',
  'September',
  'October',
  'November',
  'December',
];

export const YEARS = range(1700, getYear(new Date()) + 1, 1);
