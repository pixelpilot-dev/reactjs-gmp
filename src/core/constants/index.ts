import { range } from 'lodash';
import { getYear } from '../utils/date';

export const MONTH = [
  'January',
  'February',
  'March',
  'April',
  'May',
  'June',
  'July',
  'August',
  'September',
  'October',
  'November',
  'December',
];

export const YEARS = range(1700, getYear(new Date()) + 1, 1);

export const GENRES = [
  'Action',
  'Adventure',
  'Drama',
  'Biography',
  'Music',
  'Oscar winning Movie',
  'Documentary',
  'Comedy',
  'Horror',
  'Crime',
  'Romance',
  'Animation',
  'Family',
  'Fantasy',
  'Science Fiction',
  'History',
  'Thriller',
  'Mystery',
  'Film-Noir',
  'War',
  'Western',
  'Musical',
  'Sport',
];

export const ERROR_CAPTION_FOR_FORMS = {
  REQUIRED: 'Required',
  IS_EMAIL: 'Email is invalid',
  IS_NUMBER: 'Should only contain numbers',
  IS_MIN: 'Must be greader {0}',
  IS_MAX: 'Must be less than or equal to {0}',
  IS_URL: 'Must be url',
};

export const NOTIFICATION_TYPES = {
  SUCCESS: 'success',
  ERROR: 'error',
  WARNING: 'warning',
};
