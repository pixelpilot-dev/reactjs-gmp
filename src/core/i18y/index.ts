import en from './en';

export const LOCALE = 'EN';

export const I18Y = {
  EN: en,
};
