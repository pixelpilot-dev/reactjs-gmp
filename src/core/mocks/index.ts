export const MOVIE = {
  id: 406997,
  title: 'Wonder',
  tagline: 'Are you ready to meet Auggie Pullman?',
  vote_average: 8.2,
  vote_count: 1620,
  release_date: '2017-11-13',
  poster_path: 'https://image.tmdb.org/t/p/w500/ouYgAatYH4JzIThj6FI3UYf31RI.jpg',
  overview: 'The story of August Pullman – a boy with facial differences – who enters fifth grade, attending a mainstream elementary school for the first time.',
  budget: 20000000,
  revenue: 0,
  genres: ['Drama', 'Family'],
  runtime: 113
};

export const MOVIES = [{"id":412302,"title":"Gemini","tagline":"","vote_average":10,"vote_count":3,"release_date":"2018-03-30","poster_path":"https://image.tmdb.org/t/p/w500/oIltQs7MPk7VQFG3DJfgC63mShU.jpg","overview":"A heinous crime tests the complex relationship between a tenacious personal assistant and her Hollywood starlet boss. As the assistant travels across Los Angeles to unravel the mystery, she must stay one step ahead of a determined policeman and confront her own understanding of friendship, truth and celebrity.","budget":0,"revenue":0,"genres":["Mystery","Thriller"],"runtime":92},{"id":19404,"title":"Dilwale Dulhania Le Jayenge","tagline":"Come... Fall in Love","vote_average":9.2,"vote_count":1297,"release_date":"1995-10-20","poster_path":"https://image.tmdb.org/t/p/w500/uC6TTUhPpQCmgldGyYveKRAu8JN.jpg","overview":"Raj is a rich, carefree, happy-go-lucky second generation NRI. Simran is the daughter of Chaudhary Baldev Singh, who in spite of being an NRI is very strict about adherence to Indian values. Simran has left for India to be married to her childhood fiancé. Raj leaves for India with a mission at his hands, to claim his lady love under the noses of her whole family. Thus begins a saga.","budget":13200000,"revenue":100000000,"genres":["Comedy","Drama","Romance"],"runtime":190},{"id":425895,"title":"Tisay","tagline":"","vote_average":9,"vote_count":1,"release_date":"2016-11-14","poster_path":"https://image.tmdb.org/t/p/w500/lt03HzS8tfDg5ipUQcLqIRNs0sC.jpg","overview":"In the world of semi-pro basketball, a beautiful bookie recruits a naïve but promising player to the underworld of game-fixing.","budget":0,"revenue":0,"genres":["Drama","Romance"],"runtime":100},{"id":511679,"title":"Héctor El Father: Conocerás la verdad","tagline":"","vote_average":8.8,"vote_count":5,"release_date":"2018-03-22","poster_path":"https://image.tmdb.org/t/p/w500/jwJoURyfm4XxtnYRtIOS2pYR9Np.jpg","overview":"","budget":0,"revenue":0,"genres":["Drama"],"runtime":90},{"id":278,"title":"The Shawshank Redemption","tagline":"Fear can hold you prisoner. Hope can set you free.","vote_average":8.6,"vote_count":9868,"release_date":"1994-09-23","poster_path":"https://image.tmdb.org/t/p/w500/9O7gLzmreU0nGkIB6K3BsJbzvNv.jpg","overview":"Framed in the 1940s for the double murder of his wife and her lover, upstanding banker Andy Dufresne begins a new life at the Shawshank prison, where he puts his accounting skills to work for an amoral warden. During his long stretch in prison, Dufresne comes to be admired by the other inmates -- including an older prisoner named Red -- for his integrity and unquenchable sense of hope.","budget":25000000,"revenue":28341469,"genres":["Drama","Crime"],"runtime":142},{"id":238,"title":"The Godfather","tagline":"An offer you can't refuse.","vote_average":8.5,"vote_count":7437,"release_date":"1972-03-14","poster_path":"https://image.tmdb.org/t/p/w500/rPdtLWNsZmAtoZl9PK7S2wE3qiS.jpg","overview":"Spanning the years 1945 to 1955, a chronicle of the fictional Italian-American Corleone crime family. When organized crime family patriarch, Vito Corleone barely survives an attempt on his life, his youngest son, Michael steps in to take care of the would-be killers, launching a campaign of bloody revenge.","budget":6000000,"revenue":245066411,"genres":["Drama","Crime"],"runtime":175},{"id":372058,"title":"Your Name.","tagline":"","vote_average":8.5,"vote_count":2129,"release_date":"2016-08-26","poster_path":"https://image.tmdb.org/t/p/w500/xq1Ugd62d23K2knRUx6xxuALTZB.jpg","overview":"High schoolers Mitsuha and Taki are complete strangers living separate lives. But one night, they suddenly switch places. Mitsuha wakes up in Taki’s body, and he in hers. This bizarre occurrence continues to happen randomly, and the two must adjust their lives around each other.","budget":0,"revenue":355398372,"genres":["Romance","Animation","Drama"],"runtime":106},{"id":418827,"title":"Sister's Younger Husband","tagline":"A sexy brother-in-law approached her? And he’s a younger man?","vote_average":8.5,"vote_count":2,"release_date":"2016-10-11","poster_path":"https://image.tmdb.org/t/p/w500/wKCkyb7dyDTbZn7qy6zh8oqtFrT.jpg","overview":"A sexy brother-in-law approached her? And he’s a younger man?  Mi-yeon, who was adopted into her family when she was little, is living with her older sister, Si-yeon and Si-yeon’s husband. Mi-yeon’s desire towards her brother-in-law grows more and more every night when she gets to watch the affectionate interaction between her older sister and the brother-in-law. One day, Mi-yeon seduces her brother-in-law and the two end up crossing the line when they should not have. As the situation goes on, Mi-yeon finds out her older sister is also having affairs outside the marriage. Mi-yeon tries to clear up the relationship with her brother-in-law out of guilty feeling. However, she hears something shocking from the brother-in-law.","budget":0,"revenue":0,"genres":["Comedy","Drama","Romance"],"runtime":75},{"id":129,"title":"Spirited Away","tagline":"","vote_average":8.4,"vote_count":5095,"release_date":"2001-07-20","poster_path":"https://image.tmdb.org/t/p/w500/dL11DBPcRhWWnJcFXl9A07MrqTI.jpg","overview":"A young girl, Chihiro, becomes trapped in a strange new world of spirits. When her parents undergo a mysterious transformation, she must call upon the courage she never knew she had to free her family.","budget":15000000,"revenue":274925095,"genres":["Animation","Family","Fantasy"],"runtime":125},{"id":240,"title":"The Godfather: Part II","tagline":"I don't feel I have to wipe everybody out, Tom. Just my enemies.","vote_average":8.4,"vote_count":4308,"release_date":"1974-12-20","poster_path":"https://image.tmdb.org/t/p/w500/bVq65huQ8vHDd1a4Z37QtuyEvpA.jpg","overview":"In the continuing saga of the Corleone crime family, a young Vito Corleone grows up in Sicily and in 1910s New York. In the 1950s, Michael Corleone attempts to expand the family business into Las Vegas, Hollywood and Cuba.","budget":13000000,"revenue":102600000,"genres":["Drama","Crime"],"runtime":200},{"id":449176,"title":"Love, Simon","tagline":"He's done keeping his story straight.","vote_average":8.4,"vote_count":82,"release_date":"2018-03-16","poster_path":"https://image.tmdb.org/t/p/w500/5YUYg5q7QfC4IoNwNUtiwdiYKPr.jpg","overview":"Everyone deserves a great love story. But for seventeen-year old Simon Spier it’s a little more complicated: he’s yet to tell his family or friends he’s gay and he doesn’t actually know the identity of the anonymous classmate he’s fallen for online.","budget":17000000,"revenue":11800000,"genres":["Comedy","Drama","Romance"],"runtime":110},{"id":424,"title":"Schindler's List","tagline":"Whoever saves one life, saves the world entire.","vote_average":8.4,"vote_count":5574,"release_date":"1993-11-29","poster_path":"https://image.tmdb.org/t/p/w500/yPisjyLweCl1tbgwgtzBCNCBle.jpg","overview":"The true story of how businessman Oskar Schindler saved over a thousand Jewish lives from the Nazis while they worked as slaves in his factory during World War II.","budget":22000000,"revenue":321365567,"genres":["Drama","History","War"],"runtime":195},{"id":122,"title":"The Lord of the Rings: The Return of the King","tagline":"The eye of the enemy is moving.","vote_average":8.3,"vote_count":9857,"release_date":"2003-12-01","poster_path":"https://image.tmdb.org/t/p/w500/rCzpDGLbOoPwLjy3OAm5NUPOTrC.jpg","overview":"Aragorn is revealed as the heir to the ancient kings as he, Gandalf and the other members of the broken fellowship struggle to save Gondor from Sauron's forces. Meanwhile, Frodo and Sam bring the ring closer to the heart of Mordor, the dark lord's realm.","budget":94000000,"revenue":1118888979,"genres":["Adventure","Fantasy","Action"],"runtime":201},{"id":550,"title":"Fight Club","tagline":"Mischief. Mayhem. Soap.","vote_average":8.3,"vote_count":11996,"release_date":"1999-10-15","poster_path":"https://image.tmdb.org/t/p/w500/adw6Lq9FiC9zjYEpOqfq03ituwp.jpg","overview":"A ticking-time-bomb insomniac and a slippery soap salesman channel primal male aggression into a shocking new form of therapy. Their concept catches on, with underground \"fight clubs\" forming in every town, until an eccentric gets in the way and ignites an out-of-control spiral toward oblivion.","budget":63000000,"revenue":100853753,"genres":["Drama"],"runtime":139},{"id":510,"title":"One Flew Over the Cuckoo's Nest","tagline":"If he's crazy, what does that make you?","vote_average":8.3,"vote_count":3872,"release_date":"1975-11-18","poster_path":"https://image.tmdb.org/t/p/w500/2Sns5oMb356JNdBHgBETjIpRYy9.jpg","overview":"While serving time for insanity at a state mental hospital, implacable rabble-rouser, Randle Patrick McMurphy inspires his fellow patients to rebel against the authoritarian rule of head nurse, Mildred Ratched.","budget":3000000,"revenue":108981275,"genres":["Drama"],"runtime":133}];
