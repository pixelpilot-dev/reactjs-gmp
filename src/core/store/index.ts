import { combineReducers, configureStore, PreloadedState } from '@reduxjs/toolkit';

import { moviesApi } from './movies/api';

const rootReducer = combineReducers({
  [moviesApi.reducerPath]: moviesApi.reducer,
});

export const setupStore = (preloadedState?: PreloadedState<RootState>) => {
  return configureStore({
    reducer: rootReducer,
    middleware: getDefaultMiddleware => getDefaultMiddleware().concat(moviesApi.middleware),
    preloadedState,
  });
};

export type RootState = ReturnType<typeof rootReducer>;
export type AppStore = ReturnType<typeof setupStore>;
export type AppDispatch = AppStore['dispatch'];
