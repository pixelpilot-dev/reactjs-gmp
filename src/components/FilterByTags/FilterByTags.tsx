import React, { useEffect, useState } from 'react';
import cn from 'classnames';
import useQueryParam from '../../hooks/useQueryParam';
import { QUERY_PARAMS_BY_MOVIES } from '../../core/constants';
import { LOCALE, I18Y } from '../../core/i18y';
import { IFilterByTagsProps } from './interfaces';
import { TGenresProps } from '../../core/types/TGenresProps';

import styles from './FilterByTags.module.scss';

export const FilterByTags: React.FC<IFilterByTagsProps> = ({ options }) => {
  const CAPTION_ALL = I18Y[LOCALE].FILTER_ALL_TAG_CAPTION;
  const [tag, setTag] = useQueryParam(QUERY_PARAMS_BY_MOVIES.GENRE);
  const [active, setActive] = useState('');

  useEffect(() => {
    setActive(!tag ? CAPTION_ALL : tag);
  }, [tag]);

  const handleActiveItem = (item: string) => {
    setActive(item);
    setTag(item);
  };

  const renderItem = (item: string) => {
    const caption = item === CAPTION_ALL ? CAPTION_ALL : item;

    return (
      <button
        key={item}
        className={cn(styles.button, { [styles.isActive]: active === caption })}
        onClick={() => handleActiveItem(item)}
        type='button'
      >
        {caption}
      </button>
    );
  };

  const renderList = (items: TGenresProps) => items.map(renderItem);

  return (
    <div className={styles.filter}>
      {renderItem(CAPTION_ALL)}
      {renderList(options)}
    </div>
  );
};
