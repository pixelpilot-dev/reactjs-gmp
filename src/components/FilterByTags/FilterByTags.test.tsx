import React from 'react';
import '@testing-library/jest-dom/extend-expect';
import { BrowserRouter } from 'react-router-dom';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { GENRES_FOR_FILTER } from '../../core/constants';

import { FilterByTags } from './FilterByTags';

describe('FilterByTags component', () => {
  it('should render correctly', () => {
    const { asFragment } = render(<FilterByTags options={GENRES_FOR_FILTER} />, {
      wrapper: BrowserRouter,
    });

    const buttonAll = 1;
    const numberOfButtons = GENRES_FOR_FILTER.length + buttonAll;
    const buttons = screen.getAllByRole('button');

    expect(buttons.length).toBe(numberOfButtons);
    expect(asFragment()).toMatchSnapshot();
  });

  it('should has active tag on click', async () => {
    const user = userEvent.setup();
    render(<FilterByTags options={GENRES_FOR_FILTER} />, { wrapper: BrowserRouter });

    const buttons = screen.getAllByRole('button');
    const btn = buttons[2];

    await user.click(btn);

    expect(btn).toHaveClass('isActive');
  });
});
