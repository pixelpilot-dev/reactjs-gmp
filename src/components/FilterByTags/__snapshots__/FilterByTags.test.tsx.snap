// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterByTags component should render correctly 1`] = `
<DocumentFragment>
  <div
    class="filter"
  >
    <button
      class="button isActive"
      type="button"
    >
      All
    </button>
    <button
      class="button"
      type="button"
    >
      Fantasy
    </button>
    <button
      class="button"
      type="button"
    >
      Comedy
    </button>
    <button
      class="button"
      type="button"
    >
      Family
    </button>
    <button
      class="button"
      type="button"
    >
      Drama
    </button>
  </div>
</DocumentFragment>
`;
